kotlinx.fuzz.api/src/main/kotlin/kotlinx/fuzz/config/KFuzzConfigBuilder.kt

```diff
@@ -0,0 +1,157 @@
+            delegates.forEach { it.setFromDefault() }
+            fallbackSteps.forEach { it.invoke(configImpl) }
+            delegates.forEach { it.setFromPropertiesMap() }
+            overrideSteps.forEach { it.invoke(configImpl) }
+
+            delegates.forEach { it.validate() }
+            isBuilt = true
+            return configImpl
+        } catch (e: Throwable) {
+            throw ConfigurationException(e)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T : Any> getPropertyDelegate(propertySelector: KFuzzConfigImpl.() -> KProperty<T>): KFuzzProperty<T> {
+        val property = propertySelector(configImpl)
+
+        val kfuzzProperty = delegatesMap[property.toString()] ?: error("no KFProp found for property '${property.name}'")
+        return kfuzzProperty as KFuzzProperty<T>
+    }
+
+    /**
+     * A fuzzing config property. The value is looked up with the following priority:
+     *
+     * 1) editOverride
+     * 2) property map
+     * 3) editFallback
+     * 4) default
+     *
+     * @param default throws
+     */
+    inner class KFuzzProperty<T : Any>(
+        val name: String,
+        private val fromString: (String) -> T,
+        private val intoString: (T) -> String,
+        private val validate: (T) -> Unit,
+        private val default: T?,
+    ) {
+        private var value: T? = null
+        private val isBuilt get() = this@KFuzzConfigBuilder.isBuilt
+
+        operator fun getValue(thisRef: Any?, property: KProperty<*>): T {
+            check(value != null) { "cannot get value, config is not built yet!" }
+            return value!!
+        }
+
+        operator fun setValue(thisRef: Any?, property: KProperty<*>, value: T) {
+            check(!isBuilt) { "cannot set value, config is already built!" }
+            this.value = value
+        }
+
+        fun setFromPropertiesMap() {
+            propertiesMap[name]?.let {
+                value = fromString(it)
+            }
+        }
+
+        fun setFromDefault() {
+            default?.let {
+                value = it
+            }
+        }
+
+        fun getStringValue(): String {
+            check(isBuilt) { "cannot get string value, config is not built yet!" }
+            return intoString(value!!)
+        }
+
+        fun validate() {
+            check(value != null) { "property '$name' was not set!" }
+            validate(value!!)
+        }
+    }
+
+    inner class KFuzzConfigImpl internal constructor() : KFuzzConfig {
+        override val global = GlobalConfigImpl(this@KFuzzConfigBuilder)
+        override val target = TargetConfigImpl(this@KFuzzConfigBuilder)
+        override val coverage = CoverageConfigImpl(this@KFuzzConfigBuilder)
+
+        // TODO: we have to know which engine to use before building the config...
+        // Viable solution for future: build a stub config, check the engine, build a new config.
+        // Will need to be careful with instantiating delegates, as they will get validated.
```

Comment:
Why can't we initialize all of them? I believe that users can have settings for several engines and use one of them by specifying `engine` field. It might be unclear how to keep API independent on supported engines, but since we directly specify jazzer here it is ok (probably)

Reply:
This is what `Will need to be careful with instantiating delegates, as they will get validated.` means here: if we try to initialize them, but properties are missing, validation will fail. I guess an alternative option is to allow some properties to remain invalid

Reply:
If nobody objects, I will leave this as-is for the time being